File: src/org/perses/grammar/solidity/preprocessed/PnfSolidityParserFacade.java
/*
 * Copyright (C) 2018-2022 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.grammar.solidity.preprocessed;

import com.google.common.primitives.ImmutableIntArray;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.StringReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;
import org.perses.antlr.ParseTreeWithParser;
import org.perses.grammar.AbstractDefaultParserFacade;
import org.perses.grammar.solidity.LanguageSolidity;

public final class PnfSolidityParserFacade
    extends AbstractDefaultParserFacade<SolidityLexer, PnfSolidityParser> {

  public PnfSolidityParserFacade() {
    super(
        LanguageSolidity.INSTANCE,
        createSeparateAntlrGrammar(
            "PnfSolidity.g4", "SolidityLexer.g4", PnfSolidityParserFacade.class),
        SolidityLexer.class,
        PnfSolidityParser.class,
        ImmutableIntArray.of(SolidityLexer.Identifier));
  }

  @Override
  protected SolidityLexer createLexer(CharStream inputStream) {
    return new SolidityLexer(inputStream);
  }

  @Override
  protected PnfSolidityParser createParser(CommonTokenStream tokens) {
    return new PnfSolidityParser(tokens);
  }

  @Override
  protected ParseTree startParsing(PnfSolidityParser parser) {
    return parser.sourceUnit();
  }

  public ParseTreeWithParser parseWithOrigSolidityParser(Path file) throws IOException {
    try (BufferedReader reader = Files.newBufferedReader(file, StandardCharsets.UTF_8)) {
      return parseWithOrigSolidityParser(reader, file.toString());
    }
  }

  public ParseTreeWithParser parseWithOrigSolidityParser(String goProgram) throws IOException {
    try (BufferedReader reader = new BufferedReader(new StringReader(goProgram))) {
      return parseWithOrigSolidityParser(reader, "<in-memory>");
    }
  }

  public ParseTreeWithParser parseWithOrigSolidityParser(String goProgram, String fileName)
      throws IOException {
    try (BufferedReader reader = new BufferedReader(new StringReader(goProgram))) {
      return parseWithOrigSolidityParser(reader, fileName);
    }
  }

  private static ParseTreeWithParser parseWithOrigSolidityParser(
      BufferedReader reader, String fileName) throws IOException {
    return parseReader(
        fileName, reader, SolidityLexer::new, SolidityParser::new, SolidityParser::sourceUnit);
  }
}

File: src/org/perses/grammar/solidity/preprocessed/PnfSolidityParserFacade.kt
/*
 * Copyright (C) 2018-2026 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.grammar.solidity.preprocessed

import com.google.common.primitives.ImmutableIntArray
import org.antlr.v4.runtime.CharStream
import org.antlr.v4.runtime.CommonTokenStream
import org.perses.antlr.ParseTreeWithParser
import org.perses.grammar.AbstractParserFacade
import org.perses.grammar.solidity.LanguageSolidity
import java.io.BufferedReader
import java.io.IOException
import java.io.StringReader
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

class PnfSolidityParserFacade :
  AbstractParserFacade(
    language = LanguageSolidity,
    antlrGrammar =
      createSeparateAntlrGrammar(
        startRuleName = "sourceUnit",
        antlrParserGrammarFileName = "PnfSolidity.g4",
        antlrLexerGrammarFileName = "SolidityLexer.g4",
        PnfSolidityParserFacade::class.java,
      ),
    identifierTokenTypes =
      ImmutableIntArray.of(SolidityLexer.Identifier, SolidityLexer.YulIdentifier),
    includeAutoDetectedIdentifierTokenTypes = false,
    lexerClass = SolidityLexer::class.java,
    parserClass = PnfSolidityParser::class.java,
  ) {
  fun parseWithOrigSolidityParser(file: Path): ParseTreeWithParser {
    Files
      .newBufferedReader(file, StandardCharsets.UTF_8)
      .use { reader -> return parseWithOrigSolidityParser(reader, file.toString()) }
  }

  fun parseWithOrigSolidityParser(goProgram: String?): ParseTreeWithParser {
    BufferedReader(
      StringReader(goProgram),
    ).use { reader -> return parseWithOrigSolidityParser(reader, "<in-memory>") }
  }

  fun parseWithOrigSolidityParser(
    goProgram: String?,
    fileName: String,
  ): ParseTreeWithParser {
    BufferedReader(
      StringReader(goProgram),
    ).use { reader -> return parseWithOrigSolidityParser(reader, fileName) }
  }

  companion object {
    @Throws(IOException::class)
    private fun parseWithOrigSolidityParser(
      reader: BufferedReader,
      fileName: String,
    ): ParseTreeWithParser =
      parseReader(
        reader,
        { input: CharStream? -> SolidityLexer(input) },
        { input: CommonTokenStream? -> SolidityParser(input) },
      ) { obj: SolidityParser -> obj.sourceUnit() }
  }
}
